apps/src/code-studio/components/header/ProjectHeader.jsx

```diff
@@ -0,0 +1,130 @@
+/* globals dashboard, appOptions */
+
+import PropTypes from 'prop-types';
+import React from 'react';
+import {connect} from 'react-redux';
+
+import ProjectImport from './ProjectImport';
+import ProjectRemix from './ProjectRemix';
+import ProjectShare from './ProjectShare';
+import ProjectUpdatedAt from './ProjectUpdatedAt';
+
+import {refreshProjectName} from '../../headerRedux';
+
+class ProjectHeader extends React.Component {
+  static propTypes = {
+    projectName: PropTypes.string.isRequired,
+    refreshProjectName: PropTypes.func.isRequired
+  };
+
+  state = {
+    editName: false,
+    savingName: false
+  };
+
+  beginEdit = () => {
+    this.setState({
+      editName: true,
+      savingName: false
+    });
+  };
+
+  saveNameChange = () => {
+    const newName = this.nameChangeInput.value.trim().substr(0, 100);
+    if (newName.length === 0) {
+      return;
+    }
+
+    dashboard.project.rename(newName, () => {
+      dashboard.header.updateTimestamp();
+      this.props.refreshProjectName();
+      this.setState({
+        editName: false,
+        savingName: false
+      });
+    });
+
+    this.setState({
+      savingName: true
+    });
+  };
+
+  renderProjectName() {
+    if (this.state.editName) {
+      // Use an uncontrolled input for the "rename" operation so our UI tests
+      // can easily interface with it
```

Comment:
Will you tell me more about this? I _think _ our UI tests interface with controlled components okay in other places.  If not, that's something we should solve (but probably not in this PR).

Reply:
The specific tests I found that failed in the controlled input scenario were using the `type_into_selector` helper: https://github.com/code-dot-org/code-dot-org/blob/02a1bfbc7f4580ea36aa219150d34f473379b9b3/dashboard/test/ui/features/step_definitions/steps.rb#L549-L554

Which uses `$.change()`, which doesn't trigger the change event in a way that can be detected by React. It's possible that it could be modified to do so, but React is picky about only picking up user-triggered events and not code-triggered events, and it seemed easier to just make the component uncontrolled.

Reply:
I'm fine with punting on this, but I do think we should fix our UI tests so that they work with controlled react components. My understanding is that jquery and react use the same browser events, so it's usually just a question of which event. for example a controlled component might listen for the input event and not the change event. Next time we run into this, let's try to crack the UI test problem. I'd be happy to get pulled in on this when it comes up next.

Reply:
Hmm, my understanding was that jQuery doesn't actually trigger a real DOM event, but a simulated one that can be picked up by listeners on its own internal event handler.

See https://jsbin.com/pesanowoqe/1/edit?js,console,output for example

> Next time we run into this, let's try to crack the UI test problem. I'd be happy to get pulled in on this when it comes up next.

:+1: and likewise!